packages/flutter/lib/src/material/date_picker.dart

```diff
@@ -356,30 +380,63 @@ class _MonthPickerState extends State<MonthPicker> {
     return (endDate.year - startDate.year) * 12 + endDate.month - startDate.month;
   }
 
+  /// Add months to a month truncated date.
+  DateTime _addMonthsToMonthDate(DateTime monthDate, int monthsToAdd) {
+    return new DateTime(monthDate.year + monthsToAdd ~/ 12, monthDate.month + monthsToAdd % 12);
+  }
+
   List<Widget> _buildItems(BuildContext context, int start, int count) {
     final List<Widget> result = new List<Widget>();
-    final DateTime startDate = new DateTime(config.firstDate.year + start ~/ 12, config.firstDate.month + start % 12);
+    final DateTime startMonthDate = _addMonthsToMonthDate(config.firstDate, start);
     for (int i = 0; i < count; ++i) {
-      DateTime displayedMonth = new DateTime(startDate.year + i ~/ 12, startDate.month + i % 12);
+      DateTime monthToBuild = _addMonthsToMonthDate(startMonthDate, i);
       result.add(new DayPicker(
-        key: new ValueKey<DateTime>(displayedMonth),
+        key: new ValueKey<DateTime>(monthToBuild),
         selectedDate: config.selectedDate,
-        currentDate: _currentDate,
+        currentDate: _todayDate,
         onChanged: config.onChanged,
-        displayedMonth: displayedMonth
+        firstDate: config.firstDate,
+        lastDate: config.lastDate,
+        displayedMonth: monthToBuild
       ));
     }
     return result;
   }
 
   void _handleNextMonth() {
-    ScrollableState state = _dayPickerListKey.currentState;
-    state?.scrollTo(state.scrollOffset.round() + 1.0, duration: _kMonthScrollDuration);
+    if (!_onLastMonth) {
+      ScrollableState state = _dayPickerListKey.currentState;
+      double newPage = state.scrollOffset.round() + 1.0;
+      state?.scrollTo(newPage, duration: _kMonthScrollDuration);
+      _monthPageChanged(newPage.toInt());
```

Comment:
Ah, what's what I was thinking when I opened #7764 since it wasn't working. If it's meant to, I'll dig a bit into why it wasn't working.